doc/python-type-hinting.md

```diff
@@ -0,0 +1,85 @@
+# Python Type Hinting
+
+As a team we have committed to adding type hints throughout the Python code. This document defines our approach to type hinting, to ensure consistency in the codebase and avoid repeating discussions about how we apply typing.
+
+## Variables
+
+Specify types for variables initialised with `None`:
+
+```python
+description: Optional[str] = None
+```
+
+Specify types for empty collections:
+
+```python
+items: list[str] = []
+mappings: dict[str, int] = {}
+```
+
+In all other places, types for variables are optional.
+
+## Standard collections
+
+For standard collections, use lower case names e.g `list` rather than `List`:
+
+```python
+items: list[int] = [1]
+mappings: set[int] = {1, 2}
+```
+
+https://www.python.org/dev/peps/pep-0585/
+
+## Generic types
+
+Specify at least the top-level type parameters for all generic types:
+
+```python
+def get_objects_matching(ids: Sequence[str]) -> dict[int, dict]
+```
+
+## Optional 
+
+When something is a single type and is optional use `Optional`:
+
+```python
+def test(self, var: Optional[int]) -> None:
+```
```

Comment:
I think a point should be made between optional in which a `NoneType` value is valid, vs Optional where you have a default value. i.e a var which has a default value of 0, is optional in the sense that the caller does not need to provide a value, but it must be of one type. eg.:
```python
def test(self, increment_by: int = 0) -> None:
```
In the example above, the var `increment_by` is "optional" but must be of type int if provided, where as the `Optional[int]` should only be used if `None` is allowed, because we may evaluate like such `if var is None` or like `if var` etc.

Reply:
Good point, I have conflated optional arguments with arguments that can have the value `None`. I will reword this.

Reply:
Updated.